core/services/pg/stats.go

```diff
@@ -90,14 +90,18 @@ func NewStatsReporter(fn StatFn, lggr logger.Logger, opts ...StatsReporterOpt) *
 }
 
 func (r *StatsReporter) Start(ctx context.Context) {
+	var wg sync.WaitGroup
+
 	run := func() {
+		wg.Add(1)
+		defer wg.Done()
 		r.lggr.Debug("Starting DB stat reporter")
 		rctx, cancelFunc := context.WithCancel(ctx)
 		r.cancel = cancelFunc
 		go r.loop(rctx)
 	}
-
 	r.once.Do(run)
+	wg.Wait()
```

Comment:
I think the call to ` wg.Add(1)` needs to be moved outside the `run` func (just before `r.once.Do(run)`).  I've seen race conditions happen where it's called at the beginning of the routine being waited on, the outer function can still get passed the `wg.Wait()` line and exit before the `run` func actually gets scheduled to run (even though it's technically already been called).

Reply:
hmm. i put it inside because it's a Once. if if add is outside then the Done has to be outside as well.

as i'm writing, i'm thinking that wait group in the wrong primitive. maybe an explicit block on a channel inside the Once, that is close()'d by the spawned go routine. let me look more. 

thanks for the ideas. i wasn't convinced that wg was the right way.

Reply:
It should be fine to call `wg.Add()` outside the run function and `wg.Done()` inside, I think that's actually a fairly typical use case.  (Unless there's something special about `once.Do()` I'm missing?  

Although I'm also looking at it, and wondering whether having the `Start()` function wait for the run() function to return before returning will help fix this test.  I'm a bit fuzzy on what the underlying problem is.